File: SpringBoot-SecKill-Scala/src/main/scala/cn/edu/jxnu/seckill/dao/GoodsDao.scala
package cn.edu.jxnu.seckill.dao

import scala.language.implicitConversions
import scala.collection.JavaConversions._
import scala.collection.immutable.{ List => _, _ }
import java.util.{ List => JavaList }
import org.apache.ibatis.annotations.Mapper
import org.apache.ibatis.annotations.Select
import org.apache.ibatis.annotations.Param
import org.apache.ibatis.annotations.Update
import cn.edu.jxnu.seckill.domain.SeckillGoods
import cn.edu.jxnu.seckill.vo.GoodsVo

/**
 * 商品dao
 *
 * @author 梦境迷离.
 * @time 2018年5月19日
 * @version v1.0
 */
trait GoodsDao {

    /**
     * 查询出所有商品视图对象
     */
    @Select(Array("select g.*,mg.stock_count, mg.start_date, mg.end_date,mg.seckill_price from seckill_goods mg left join goods g on mg.goods_id = g.id"))
    def listGoodsVo(): JavaList[GoodsVo]

    /**
     * 根据商品id=秒杀的商品id,查询出商品视图对象
     */
    @Select(Array("select g.*,mg.stock_count, mg.start_date, mg.end_date,mg.seckill_price from seckill_goods mg left join goods g on mg.goods_id = g.id where g.id = #{goodsId}"))
    def getGoodsVoByGoodsId(@Param("goodsId") goodsId: Long): GoodsVo

    /**
     * 库存减1
     */
    @Update(Array("update seckill_goods set stock_count = stock_count - 1 where goods_id = #{goodsId} and stock_count > 0"))
    def reduceStock(g: SeckillGoods): Integer

    /**
     * 库存修改
     */
    @Update(Array("update seckill_goods set stock_count = #{stockCount} where goods_id = #{goodsId}"))
    def resetStock(g: SeckillGoods): Integer

}

File: seckill/src/main/java/cn/edu/jxnu/seckill/dao/GoodsDao.java
package cn.edu.jxnu.seckill.dao;

import java.util.List;

import org.apache.ibatis.annotations.Mapper;
import org.apache.ibatis.annotations.Param;
import org.apache.ibatis.annotations.Select;
import org.apache.ibatis.annotations.Update;

import cn.edu.jxnu.seckill.domain.SeckillGoods;
import cn.edu.jxnu.seckill.vo.GoodsVO;

@Mapper
public interface GoodsDao {

    @Select("select g.*, sg.seckill_price, sg.stock_count, sg.start_time, sg.end_time from seckill_goods sg left join goods g on sg.goods_id=g.id")
    List<GoodsVO> listGoodsVO();

    @Select("select g.*, sg.seckill_price, sg.stock_count, sg.start_time, sg.end_time from seckill_goods sg left join goods g on sg.goods_id=g.id where g.id=#{goodsId}")
    GoodsVO getGoodsVOByGoodsId(@Param("goodsId") long goodsId);

    @Update("update seckill_goods set stock_count=stock_count-1 where goods_id=#{goodsId} and stock_count > 0")
    int reduceStock(SeckillGoods g);

    @Update("update seckill_goods set stock_count = #{stockCount} where goods_id = #{goodsId}")
    int resetStock(SeckillGoods g);
}
